### ISSUE TITLE
Guidance needed on the interaction between ValueGenerated.OnAdd and DefaultValue

### ISSUE THREAD

**Issue by roji** [dotnet/efcore#5366]
While working on Npgsql reverse engineering I came across something I'm having trouble understanding.

If I understand the [the docs](http://ef.readthedocs.io/en/latest/modeling/generated-properties.html) correctly, setting ValueGenerated.OnAdd on a column means that its value will be generated if the inserted entity has a default value for that field (i.e. null, 0).

My first question is whether setting DefaultValue (or DefaultValueSql) should logically also imply ValueGenerated.OnAdd. In other words, does it make any sense (say, for a reveng provider) to set a DefaultValue without setting ValueGenerated.OnAdd? If so, when would that DefaultValue actually come into play, considering that without ValueGenerated.OnAdd the CLR field's value is always taken?

If DefaultValue logically implies ValueGenerated.OnAdd, shouldn't this actually be the case? In other words, shouldn't setting DefaultValue make it so that accessing a field's value should return ValueGenerated.OnAdd (let's abstract away OnAddOrUpdate for just a moment)? I'm asking this because my reveng provider only set DefaultValue, and this resulted in a code model which doesn't include any ValueGenerated value (and for primary keys actually sets ValueGenerated.Never).

If the above doesn't make sense for some reason, should I just be setting ValueGenerated.OnAdd every time I also set a DefaultValue?

**Comment by lajones** [comment]
Hi Shay

I think this is a mix-up between the meanings of “default value” within C# and within the DB. ValueGeneratedOnAdd() and Default Value() are different things.

ValueGeneratedOnAdd() was designed with the concept of “the system” automatically generating a value for you at INSERT time e.g. for SQL Server the concept of an IDENTITY column. Within C# the user should _not_ assign any value to this Property. Instead it should let “the system” automatically assign one. Now some databases don’t have the concept of IDENTITY columns and even SQL Server can set up IDENTITY columns only for some data types – for those other kinds of Properties “the system” may in fact be you, the model developer, assigning a ValueGenerator to that column which will generate the value during the pipeline before the actual call to INSERT on the DB takes place but nonetheless “the system” is responsible for generating that value.

However, from the docs, if you _do_ assign a value to that Property in C# and then attempt to INSERT that value into the database, then we will attempt to INSERT that value (I’m not totally sure, but I don’t think this will work on e.g. SQL Server where the underlying column is an IDENTITY column – I think you’ll get an Exception. It may work on other databases but really you just shouldn’t be doing it).

DefaultValue() was designed with the concept of a DEFAULT value _on the database_ in mind e.g. SQL Server DEFAULT() statements. In this case it’s perfectly fine for you to assign a C# value to a Property and that value will get stored on the DB at INSERT time. However, if you choose not to assign a value, then the INSERT statement sent to the database will not include that value and the database will assign the DEFAULT value that was referenced in the CREATE statement for that column.

The confusion arises (I think) by how we’re telling whether you’ve “assigned a value” to a given C# Property when you add its containing Entity to the ChangeTracker and call SaveChanges() i.e. when you trigger an INSERT. As you say, according to the docs, we tell whether you’ve assigned a value to a given Property by comparing its value to the C# default for that data type – so null for nullable types, 0 for integer types… and proceed as above. But that has nothing to do with comparing that value to the database DEFAULT value.

Now Reverse Engineering (we now call it “Scaffolding”) is coming the other way: it looks at the database to see what’s present there and generates code accordingly. This depends on the provider of course but for SQL Server if it’s an IDENTITY column you’ll get ValueGeneratedOnAdd(), if there’s a DEFAULT value then the generated code will call the DefaultValue() API.

On top of this there is a Migrations Convention that if a single Property is the key for an EntityType and if that Property is a C# integer type (short, int, long…) then we will automatically assume that that Property is ValueGeneratedOnAdd(). So, when doing Scaffolding for e.g. SQL Server, in order to accommodate that Convention, if we look at a key column and it has a database DEFAULT value (and therefore is _not_ an IDENTITY column since SQL Server itself prevents you creating a column which is both) we explicitly set the ValueGeneratedOnAdd to ValueGenerated.Never so as to override the Convention.

Hope that explains everything.

Cheers,

laj

**Comment by roji** [comment]
@lajones thanks for the patient explanation! Here are a few follow-up questions to make sure I understand...

So you're saying that ValueGeneratedOnAdd() means that a value is *always* generated by the system (be it the database with SqlServer IDENTITY or the model developer); in other words that it is illegal for the user to assign a value to it before insertion. In contrast, DefaultValue() means that a value *may* be generated by the system, but only if the user hasn't explicitly set the column (to a value other than the CLR default for that type). Is that correct?

Concretely, PostgreSQL doesn't have an IDENTITY type like SqlServer - all auto-increment fields (keys or otherwise) are managed via sequences, in other words have a DEFAULT() clause getting the next value of the sequence. So if I understand correctly, Npgsql should never scaffold a model with .ValueGeneratedOnAdd(), and the proper reverse-engineering for a sequence-driven entity key would be something like:

```c#
entity.Property(e => e.Id)
    .HasDefaultValueSql("nextval('my_sequence_name')")
    .ValueGeneratedNever();
```

If I'm understanding things properly, I'll just say that it's somewhat confusing for .ValueGeneratedNever() here to mean "a value actually *will* be generated, but only if you leave this field at its CLR default" - it may be worthwhile to make the docs a bit more explicit on all this.

**Comment by roji** [comment]
One more note: I'm quite confused looking at the test SqlServerValueGenerationScenariosTest.IdentityColumnWithDefaultValue.Insert_with_default_value_from_sequence. First, the test name suggests that it *is* possible to have IDENTITY columns with default (i.e. a sequence-driven IDENTITY?). I can also see the following config:

```c#
modelBuilder
    .Entity<Blog>()
    .Property(e => e.Id)
    .HasDefaultValueSql("next value for MySequence");
```

which does not have the .ValueGeneratedNever() you mentioned, so I'm even less clear on what it means to have that or not...

**Comment by ajcvickers** [comment]
@roji I have a slightly different perspective from @lajones. ValueGenerated.OnAdd does mean that a value will be generated if the value set is the CLR default for the property type. This is a core concept, rather than a relational concept--how that value gets generated is not specified at all. In relational code setting some default value for a column almost always means that the property should be marked as OnAdd, with the additional information to relational code on how the value will be obtained. This is why the fluent API for setting, for example, DefaultValueSql also sets OnAdd:
```C#
public static PropertyBuilder HasDefaultValueSql(
    [NotNull] this PropertyBuilder propertyBuilder,
    [CanBeNull] string sql)
{
    Check.NotNull(propertyBuilder, nameof(propertyBuilder));
    Check.NullButNotEmpty(sql, nameof(sql));

    var property = (Property)propertyBuilder.Metadata;
    if (ConfigurationSource.Convention.Overrides(property.GetValueGeneratedConfigurationSource()))
    {
        property.SetValueGenerated(ValueGenerated.OnAdd, ConfigurationSource.Convention);
    }

    propertyBuilder.GetInfrastructure<InternalPropertyBuilder>().Relational(ConfigurationSource.Explicit).HasDefaultValueSql(sql);

    return propertyBuilder;
}
```
There are some situations where you may want a default value but have a computed column, which is why OnAdd is only set if it hasn't already been explicitly set to something else.

So I expect all the code you have is fine, except that in your fluent API for setting the default value you should do something similar to the code above to also set OnAdd.

**Comment by roji** [comment]
@ajcvickers thanks for your response. What you makes sense to me, but I can see some code that may not correspond. Specifically, in SqlServerScaffoldingModelFactory.VisitPrimaryKey I can see a check whether the key is an identity. If not, and it's an integer or guid, ValueGenerated.Never is added. If I understand correctly, that means that sequence-driven integer keys (not IDENTITY) are scaffolded with ValueGenerated.Never although they clearly have a default...

Is this simply a bug in the scaffolding provider, or something deeper? It seems to me that there may be more than one meaning attached to ValueGenerated in different parts of the code..

**Comment by ajcvickers** [comment]
@roji If scaffolding is able to determine that the reason the key column has a default is because that default will generate a key value (which seems reasonable but I'm not sure how much can be inferred from the store metadata available) then it seems to me like this is a bug in scaffolding, or, at least, something that is currently not supported by scaffolding.

**Comment by lajones** [comment]
@roji I've done some investigation on this. I still need to do some more. But the short answer is, after discussing with @divega, I agree with @ajcvickers and @divega - I think you can just call `HasDefaultValueSql()`, it will automatically set the property to `ValueGenerated.OnAdd` and you should be good to go.

**The long answer**

Re the test `SqlServerValueGenerationScenariosTest.IdentityColumnWithDefaultValue.Insert_with_default_value_from_sequence`. The test is misnamed. Strictly speaking the column in question is not an IDENTITY column. Instead it's a column with a sequence. Now that could be looked on as being fussy about semantics; both IDENTITY and SEQUENCE do very similar things - there are pluses and minuses to both (see e.g. [this] (http://stackoverflow.com/questions/10062328/sequence-vs-identity)).

Nonetheless the way you define an "IDENTITY column" and a "column with a sequence" are different SQL Server CREATE statements, SQL Server stores different metadata about the 2 types of columns, and SQL Server _does_ prevent you from simultaneously defining a column as an IDENTITY column and at the same time assigning a DEFAULT(...) statement to it. You get the error:

`Defaults cannot be created on columns with an IDENTITY attribute`

Also IDENTITY columns will error if you try to insert a value into them whereas columns with a sequence will not - they will just accept the value you choose. So if the table underlying your Entity has a column with IDENTITY set on it _and_ you choose to try to assign a value to that column (by having the Property set to a non-CLR-default value) then expect an Exception.

We should probably just rename the test.

Re the rest. I had a chat with @divega and it seems that if you set `HasDefaultValueSql()` then you should automatically have  `ValueGenerated.OnAdd` set on the same property (which is what @ajcvickers was saying above). According to the XML docs `ValueGenerated.OnAdd` specifies that the value is to be generated _by the database_ at INSERT time only (as opposed to at INSERT and UPDATE) and all default values fall under that heading. But since calling `HasDefaultValueSql()` sets `ValueGenerated.OnAdd` anyway you're good to go.

He also explained that, contrary to what I said originally, there are good reasons when a user might wish to override a Property designated as `ValueGenerated.OnAdd` with a given non-default value from C# at INSERT time - examples are a) if you're seeding data to a database and (temporarily) have `IDENTITY INSERT` set on, or b) if you're inserting into a "column with a sequence" but you don't want to use the sequence for this particular entry.

I still need more data on how we distinguish between values which are generated _on the database_ at INSERT time and values which are generated _on the client_ at INSERT time. I'm not sure whether the latter needs `ValueGenerated.OnAdd` also - @ajcvickers can comment on that but from a Scaffolding POV you can't know what is intended to be generated on the client and so don't need to know.

The `ValueGenerated.Never` issue is kind of separate but you're right - I think there's a bug there.

If the key to a table is a single column and the type would be mapped to a C# integer type or a Guid then we have a Convention that will automatically decide that that column is `ValueGenerated.OnAdd`. During Scaffolding, if the column is just a normal column with no kind of IDENTITY or sequence attached to it, but it happens to be a single key column of integer or Guid type then we need to assign `ValueGenerated.Never` to override the Convention. However we need to _not_ do this if the column's value would be automatically created by the database. IDENTITY is one way of doing that and we do correctly do _not_ assign `ValueGenerated.Never` under those circumstances. However, as you point out, if the single key column is generated via a default value (whether that default value references a sequence or not) then that is another scenario where the database is generating the value. Currently we do not correctly identify that scenario and incorrectly we do assign `ValueGenerated.Never`. We need to update the code at https://github.com/aspnet/EntityFramework/blob/dev/src/Microsoft.EntityFrameworkCore.SqlServer.Design/Internal/SqlServerScaffoldingModelFactory.cs#L92 to also check whether the column has a default value and if so _not_ assign `ValueGenerated.Never`. Thanks for the help in finding that issue.

**Comment by roji** [comment]
@lajones and @ajcvickers, thanks again for the patient explanations, I think everything is good now.

Things that need to be done:

* To properly support scaffolding of sequence-driven keys in the Npgsql provider, I have already changed [the code here](https://github.com/aspnet/EntityFramework/blob/dev/src/Microsoft.EntityFrameworkCore.SqlServer.Design/Internal/SqlServerScaffoldingModelFactory.cs#L103) from:

```c#
if (propertyType?.IsIntegerForSerial() == true || propertyType == typeof(Guid))
```

to:

```c#
if ((propertyType?.IsIntegerForSerial() == true || propertyType == typeof(Guid)) &&
    !pkColumns[0].ValueGenerated.HasValue)
```

In other words, we set ValueGenerated.Never on a key only if it doesn't have a default defined. I suggest you guys make the same change in your own providers.

I agree that the IdentityColumnWithDefaultValue test seems to be just a confusing naming issue.

Thanks again for the guidance!
